// react-docgen-displayname-handler/distribution/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createDisplayNameHandler = createDisplayNameHandler;

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _recast = require('recast');

var _recast2 = _interopRequireDefault(_recast);

var _reactDocgen = require('react-docgen');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var getMemberValuePath = _reactDocgen.utils.getMemberValuePath,
    getNameOrValue = _reactDocgen.utils.getNameOrValue,
    isExportsOrModuleAssignment = _reactDocgen.utils.isExportsOrModuleAssignment;
var types = _recast2.default.types.namedTypes;


var DEFAULT_NAME = 'UnknownComponent';

function getNameFromPath(path) {
  var node = path.node;
  switch (node.type) {
    case types.Identifier.name:
    case types.Literal.name:
      return getNameOrValue(path);
    case types.MemberExpression.name:
      return _reactDocgen.utils.getMembers(path).reduce(function (name, _ref) {
        var path = _ref.path,
            computed = _ref.computed;
        return computed && getNameFromPath(path) ? name : `${name}.${getNameFromPath(path) || ''}`;
      }, getNameFromPath(path.get('object')));
    default:
      return null;
  }
}

function getStaticDisplayName(path) {
  var displayName = null;
  var staticMember = getMemberValuePath(path, 'displayName');
  if (staticMember && types.Literal.check(staticMember.node)) {
    displayName = getNameFromPath(staticMember);
  }

  return displayName || null;
}

function getNodeIdentifier(path) {
  var displayName = null;
  if (types.FunctionExpression.check(path.node) || types.FunctionDeclaration.check(path.node) || types.ClassExpression.check(path.node) || types.ClassDeclaration.check(path.node)) {
    displayName = getNameFromPath(path.get('id'));
  }

  return displayName || null;
}

function getVariableIdentifier(path) {
  var displayName = null;
  var searchPath = path;

  while (searchPath !== null) {
    if (types.VariableDeclarator.check(searchPath.node)) {
      displayName = getNameFromPath(searchPath.get('id'));
      break;
    }
    if (types.AssignmentExpression.check(searchPath.node) && !isExportsOrModuleAssignment(searchPath)) {
      displayName = getNameFromPath(searchPath.get('left'));
      break;
    }
    searchPath = searchPath.parentPath;
  }

  return displayName || null;
}

function getNameFromFilePath() {
  var filePath = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';

  var displayName = null;

  var filename = _path2.default.basename(filePath, _path2.default.extname(filePath));
  if (filename === 'index') {
    var parts = _path2.default.dirname(filePath).split(_path2.default.sep);
    displayName = parts[parts.length - 1];
  } else {
    displayName = filename;
  }

  return displayName.charAt(0).toUpperCase().concat(displayName.slice(1)).replace(/-([a-z])/, function (_, match) {
    return match.toUpperCase();
  });
}

function createDisplayNameHandler(filePath) {
  return function displayNameHandler(documentation, path) {
    var displayName = [getStaticDisplayName, getNodeIdentifier, getVariableIdentifier].reduce(function (name, getDisplayName) {
      return name || getDisplayName(path);
    }, '');

    if (!displayName) {
      displayName = getNameFromFilePath(filePath);
    }

    documentation.set('displayName', displayName || DEFAULT_NAME);
  };
}

exports.default = createDisplayNameHandler('');
